[improve][io] Fix Kotlin version issue for functions built in Kotlin

abhilashmandaliya opened the pull request:
### Motivation

I was trying to implement a custom Sink connector using Kotlin. My Kotlin version was 1.8.10 and was using Jackson in my code. I saw a weird reflection-related error in my code. The debugging ended up with a finding that the Pulsar broker was bringing the Kotlin library dependency which was version 1.4.32 and that was getting loaded before my version. Hence my sink was using the older version of a specific class. While looking for Kotlin support in Pulsar code and feedback from the committers in this PR, I realized that Pulsar code is not using the Kotlin library directly but it is a transitive dependency. The latest stable version of the main dependency(okhttp3) bringing Kotlin(1.6.20) as a transitive dependency doesn't use the latest version of Kotlin(1.8.10). Hence, updated `okhttp3`, `okio`, and `Kotlin` separately for the convenience of an existing structure. 

Putting the exception that I got here. (Not sure whether that is required):

```
Caused by: kotlin.reflect.jvm.internal.KotlinReflectionInternalError: Cannot create type for an unsupported classifier: class kotlin.Unit (Kotlin reflection is not available) (class kotlin.jvm.internal.ClassReference)
        at kotlin.reflect.full.KClassifiers.createType(KClassifiers.kt:48) ~[kotlin-reflect-1.8.10.jar:1.8.10-release-430(1.8.10)]
        at kotlin.reflect.full.KClassifiers.createType$default(KClassifiers.kt:42) ~[kotlin-reflect-1.8.10.jar:1.8.10-release-430(1.8.10)]
        at com.fasterxml.jackson.module.kotlin.KotlinAnnotationIntrospector.<clinit>(KotlinAnnotationIntrospector.kt:218) ~[jackson-module-kotlin-2.14.2.jar:2.14.2]
        at com.fasterxml.jackson.module.kotlin.KotlinModule.setupModule(KotlinModule.kt:121) ~[jackson-module-kotlin-2.14.2.jar:2.14.2]
        at com.fasterxml.jackson.databind.ObjectMapper.registerModule(ObjectMapper.java:853) ~[com.fasterxml.jackson.core-jackson-databind-2.13.4.2.jar:2.13.4.2]
```
Here `kotlin.Unit` class is loaded from `Kotlin 1.4.32` whereas Jackson was compiled with `Kotlin 1.5.32`.

### Modifications

* Upgraded transitive `org.jetbrains.kotlin:kotlin-stdlib` library from `1.4.32` to `1.8.10`
* Upgraded transitive `okhttp3` version from `4.9.3` to `4.10.0`
* Upgraded transitive `okio` version from `2.8.0` to `3.0.0`

### Verifying this change

- [x] Make sure that the change passes the CI checks.

*(Please pick either of the following options)*

This change is a trivial rework.

### Does this pull request potentially affect one of the following parts:

<!-- DO NOT REMOVE THIS SECTION. CHECK THE PROPER BOX ONLY. -->

*If the box was checked, please highlight the changes*

- [x] Dependencies (add or upgrade a dependency)
- [ ] The public API
- [ ] The schema
- [ ] The default values of configurations
- [ ] The threading model
- [ ] The binary protocol
- [ ] The REST endpoints
- [ ] The admin CLI options
- [ ] The metrics
- [ ] Anything that affects deployment

### Documentation

<!-- DO NOT REMOVE THIS SECTION. CHECK THE PROPER BOX ONLY. -->

- [ ] `doc` <!-- Your PR contains doc changes. -->
- [ ] `doc-required` <!-- Your PR changes impact docs and you will update later -->
- [x] `doc-not-needed` <!-- Your PR changes do not impact docs -->
- [ ] `doc-complete` <!-- Docs have been already added -->

### Matching PR in forked repository

PR in forked repository: [https://github.com/abhilashmandaliya/pulsar/pull/1](https://github.com/abhilashmandaliya/pulsar/pull/1)

<!--
After opening this PR, the build in apache/pulsar will fail and instructions will
be provided for opening a PR in the PR author's forked repository.

apache/pulsar pull requests should be first tested in your own fork since the 
apache/pulsar CI based on GitHub Actions has constrained resources and quota.
GitHub Actions provides separate quota for pull requests that are executed in 
a forked repository.

The tests will be run in the forked repository until all PR review comments have
been handled, the tests pass and the PR is approved by a reviewer.
-->

nodece commented on pom.xml:
```diff
@@ -1966,8 +1965,6 @@ flexible messaging model and an intuitive client API.</description>
         <test.additional.args/>
         <maven.compiler.source>8</maven.compiler.source>
         <maven.compiler.target>8</maven.compiler.target>
-        <pulsar.broker.compiler.release></pulsar.broker.compiler.release>
```
This should be done intentionally.

tisonkun replied:
@abhilashmandaliya to fix IDE warning, you can collapse the tag instead of removing it.

Anyway, keep one PR for one purpose (upgrade kotlin lib) can help you get a smooth review experience - we don't distract attention.

abhilashmandaliya replied:
Thanks, @nodece and @tisonkun.  Have reverted that change.

tisonkun reviewed (changes requested):
I don't think it's an isolated version issue. Instead, you should take a look of how Pulsar uses okio and make an upgrade from the direct deps and then propagate to transitive deps.

tisonkun commented:
cc @nicoloboschi since you're the last author for editing these versions.

abhilashmandaliya commented:
@nodece @tisonkun @lhotari @nicoloboschi 
I have made the changes as per the feedback and also updated the PR description.
GitHub UI did some weird things hence there are several reviews and remove status lines.

nicoloboschi commented on pom.xml:
we should move to 3.3.0, based on the [changelog](https://github.com/square/okio/blob/master/CHANGELOG.md) it's more stable and it upgrades kotlin to 1.8 as well

abhilashmandaliya commented:
@nicoloboschi PTAL at the updated library version change.

nicoloboschi reviewed (approved):
LGTM

tisonkun commented:
@abhilashmandaliya license check failed:

```
com.squareup.okio-okio-jvm-3.0.0.jar unaccounted for in LICENSE
org.jetbrains.kotlin-kotlin-stdlib-1.6.20.jar unaccounted for in LICENSE
org.jetbrains.kotlin-kotlin-stdlib-common-1.5.31.jar unaccounted for in LICENSE
org.jetbrains.kotlin-kotlin-stdlib-jdk7-1.6.10.jar unaccounted for in LICENSE
org.jetbrains.kotlin-kotlin-stdlib-jdk8-1.6.10.jar unaccounted for in LICENSE
com.squareup.okio-okio-3.3.0.jar mentioned in LICENSE, but not bundled
org.jetbrains.kotlin-kotlin-stdlib-1.8.0.jar mentioned in LICENSE, but not bundled
org.jetbrains.kotlin-kotlin-stdlib-common-1.8.0.jar mentioned in LICENSE, but not bundled
org.jetbrains.kotlin-kotlin-stdlib-jdk7-1.8.0.jar mentioned in LICENSE, but not bundled
org.jetbrains.kotlin-kotlin-stdlib-jdk8-1.8.0.jar mentioned in LICENSE, but not bundled
```

The first lines can be redundant that you should remove. The final lines about kotlin-stdlib, I'm afraid, indicate an issue that we are missing to bundle them?

abhilashmandaliya commented:
> @abhilashmandaliya license check failed:
> 
> ```
> com.squareup.okio-okio-jvm-3.0.0.jar unaccounted for in LICENSE
> org.jetbrains.kotlin-kotlin-stdlib-1.6.20.jar unaccounted for in LICENSE
> org.jetbrains.kotlin-kotlin-stdlib-common-1.5.31.jar unaccounted for in LICENSE
> org.jetbrains.kotlin-kotlin-stdlib-jdk7-1.6.10.jar unaccounted for in LICENSE
> org.jetbrains.kotlin-kotlin-stdlib-jdk8-1.6.10.jar unaccounted for in LICENSE
> com.squareup.okio-okio-3.3.0.jar mentioned in LICENSE, but not bundled
> org.jetbrains.kotlin-kotlin-stdlib-1.8.0.jar mentioned in LICENSE, but not bundled
> org.jetbrains.kotlin-kotlin-stdlib-common-1.8.0.jar mentioned in LICENSE, but not bundled
> org.jetbrains.kotlin-kotlin-stdlib-jdk7-1.8.0.jar mentioned in LICENSE, but not bundled
> org.jetbrains.kotlin-kotlin-stdlib-jdk8-1.8.0.jar mentioned in LICENSE, but not bundled
> ```
> 
> The first lines can be redundant that you should remove. The final lines about kotlin-stdlib, I'm afraid, indicate an issue that we are missing to bundle them?

@tisonkun With the latest okio library, we don't need to use a different version of Kotlin hence I removed it from the `pom.xml`. I forgot to remove it from the LICENCE file. I have fixed that. Not sure why okio error is there.

tisonkun commented:
@abhilashmandaliya I get the error report wrong.

```
com.squareup.okio-okio-jvm-3.0.0.jar unaccounted for in LICENSE
org.jetbrains.kotlin-kotlin-stdlib-1.6.20.jar unaccounted for in LICENSE
org.jetbrains.kotlin-kotlin-stdlib-common-1.5.31.jar unaccounted for in LICENSE
org.jetbrains.kotlin-kotlin-stdlib-jdk7-1.6.10.jar unaccounted for in LICENSE
org.jetbrains.kotlin-kotlin-stdlib-jdk8-1.6.10.jar unaccounted for in LICENSE
```

These lines mean that you're now (transitive) depending on these libs but they are not listed (mentioned) in the LICENSE file.

You may double check if the version is correct (kotlin 1.5 and 1.6 while we plan to use 1.8? okio-jvm 3.0.0?). The LICENSE updating job I can do you a favor. But we should confirm the final status logically first.

tisonkun commented on pom.xml:
```diff
@@ -1326,40 +1324,53 @@ flexible messaging model and an intuitive client API.</description>
         <groupId>com.squareup.okhttp3</groupId>
         <artifactId>logging-interceptor</artifactId>
         <version>${okhttp3.version}</version>
+        <exclusions>
```
@abhilashmandaliya if we exclude kotlin dep here, is it possible we are missing bundling several jars so that the function shade jar won't work?

tisonkun reviewed (changes requested):
I push a commit to suppress:

```
org.jetbrains.kotlin-kotlin-stdlib-1.8.20.jar mentioned in LICENSE, but not bundled
org.jetbrains.kotlin-kotlin-stdlib-common-1.8.20.jar mentioned in LICENSE, but not bundled
org.jetbrains.kotlin-kotlin-stdlib-jdk7-1.8.20.jar mentioned in LICENSE, but not bundled
org.jetbrains.kotlin-kotlin-stdlib-jdk8-1.8.20.jar mentioned in LICENSE, but not bundled
org.jetbrains-annotations-13.0.jar mentioned in LICENSE, but not bundled
```

But then the changeset seems a bit strange. Comment inline.

Also, you can ping the reviewers when you make progress - pushing new commits doesn't send notification.

abhilashmandaliya commented:
Closing the PR as Pulsar 3.0.0 doesn't bring the older Kotlin version. Thanks a lot, @tisonkun for your help.
